drivers/periph: renamed random and more

haukepetersen opened the pull request:
The random peripheral driver interface needed an overhaul to comply with the other modernized peripheral interfaces:
- renamed it to 'hrng' -> many people complained about the old name...
- removed power on/off functions (as the device should be powered off by default and only powered on while used)
- none of our implementations have any valid error case, so no need for a return value on 'read'

This also fixes #3109

Travis won't be happy, as the path for the `test/pkg_micro-ecc` is still untouched.. .Would somebody mind to tell me the fasted way to change that patch?

gebart commented:
How about hwrng instead of hrng? I think hwrng makes it more obvious what this is to the uninitiated reader.

gebart commented on tests/pkg_micro-ecc/main.c:
```diff
@@ -61,9 +61,9 @@ int main(void)
+    hrng_init();
     /* power off RNG to save energy */
-    random_poweroff();
+    hrng_poweroff();
```
poweroff was removed, was it not?

gebart commented:
I agree with the general idea, `random` is too generic and should be changed :+1:

kaspar030 commented:
> I agree with the general idea, random is too generic and should be changed :+1:

:+1:

Or maybe just rng?

gebart commented:
@kaspar030 doesn't that make it a bit confusing since we have software solutions prefixed with prng?

haukepetersen commented:
Actually, i would also prefer `hwrng` not that I think about it. @kaspar030, are you okay with that, then I will change it.

kaspar030 commented:
Fine with me!

haukepetersen commented:
how about this? let me know when I can squash. And also help with adapting this stupid patch file for `pkg/micro-ecc` would be appreciated.

kaspar030 commented on boards/mulle/include/periph_conf.h:
```diff
@@ -505,9 +505,8 @@ extern "C"
-#define RANDOM_NUMOF            (1U)
-#define RANDOM_CLKEN()          (BITBAND_REG32(SIM->SCGC3, SIM_SCGC3_RNGA_SHIFT) = 1)
-#define RANDOM_CLKDIS()         (BITBAND_REG32(SIM->SCGC3, SIM_SCGC3_RNGA_SHIFT) = 0)
+#define HRNG_CLKEN()        (BITBAND_REG32(SIM->SCGC3, SIM_SCGC3_RNGA_SHIFT) = 1)
```
HWRNG

haukepetersen replied:
don't care, its a CPU internally function... But if you insist...

kaspar030 commented on boards/pba-d-01-kw2x/include/periph_conf.h:
```diff
@@ -325,10 +325,9 @@ extern "C"
-#define RANDOM_CLKEN()       (SIM->SCGC6 |= (1 << 9))
-#define RANDOM_CLKDIS()      (SIM->SCGC6 &= ~(1 << 9))
+#define KINETIS_RNGA        RNG
+#define HRNG_CLKEN()        (SIM->SCGC6 |= (1 << 9))
```
HWRNG

haukepetersen replied:
same

kaspar030 commented on cpu/kinetis_common/hwrng_rnga.c:
```diff
@@ -13,7 +13,7 @@
  *
  * @file
- * @brief       Low-level random number generator driver implementation.
+ * @brief       HRNG interface implementation
```
here

kaspar030 commented on cpu/kinetis_common/hwrng_rnga.c:
```diff
@@ -22,21 +22,24 @@
     unsigned int count = 0;
 
+    /* power on and enable the device */
+    HRNG_CLKEN();
```
here

kaspar030 commented on cpu/kinetis_common/hwrng_rnga.c:
```diff
@@ -50,31 +53,14 @@ int random_read(char *buf, unsigned int num)
-/*
-void isr_rng(void)
-{
+    HRNG_CLKDIS();
```
here

kaspar030 commented on cpu/kinetis_common/hwrng_rngb.c:
```diff
@@ -13,32 +13,42 @@
 {
     unsigned int count = 0;
 
+    HRNG_CLKEN();
```
...

kaspar030 commented:
You forgot the capital HRNG's.

gebart commented:
@haukepetersen make another patch on top of the 0002-Include-RIOT-Hardware-RNG-interface.patch
make your changes in the checked-out copy of micro-ecc and use git format-patch to create the patch file, then add it as 0003-RIOT-hwrng-API-changes.patch
The two patches can also be merged into one if desired, just use `git diff HEAD~2 > merged.patch`

I'm assuming the actual change is not much more than a search and replace random_ to hwrng_, and removing the random_poweron/off calls?

haukepetersen commented:
Left them out because they are cpu internal, but I will fix it to make it look nicer...

haukepetersen commented:
@gebart: yepp, just stome string replace and remove operations. Will do.

haukepetersen commented:
@kaspar030: fixed.

haukepetersen commented:
okay, this should be it. If you guys are happy I would squash and give Travis a run at it

kaspar030 commented:
pls squash!

haukepetersen commented:
squashed

gebart commented:
the commit messages need updating hrng->hwrng

gebart commented:
Could you use `format-patch -w` for pkg/micro-ecc/0003-Adapted-to-changed-random-interface.patch to get rid of all the empty lines being stripped of whitespace? it makes the actual patch content less readable.

haukepetersen commented:
yep, will do both things

gebart commented:
nucleo-f401 fails:
```
default:nucleo-f401:
Building application "default" for "nucleo-f401" with MCU "stm32f4".
/home/travis/build/RIOT-OS/RIOT/cpu/stm32f4/periph/hwrng.c: In function 'hwrng_read':
/home/travis/build/RIOT-OS/RIOT/cpu/stm32f4/periph/hwrng.c:34:21: error: 'RCC_AHB2ENR_RNGEN' undeclared (first use in this function)
     RCC->AHB2ENR |= RCC_AHB2ENR_RNGEN;
                     ^
/home/travis/build/RIOT-OS/RIOT/cpu/stm32f4/periph/hwrng.c:34:21: note: each undeclared identifier is reported only once for each function it appears in
/home/travis/build/RIOT-OS/RIOT/cpu/stm32f4/periph/hwrng.c:35:5: error: 'RNG' undeclared (first use in this function)
     RNG->CR = RNG_CR_RNGEN;
     ^
/home/travis/build/RIOT-OS/RIOT/cpu/stm32f4/periph/hwrng.c:35:15: error: 'RNG_CR_RNGEN' undeclared (first use in this function)
     RNG->CR = RNG_CR_RNGEN;
               ^
/home/travis/build/RIOT-OS/RIOT/cpu/stm32f4/periph/hwrng.c:40:28: error: 'RNG_SR_DRDY' undeclared (first use in this function)
         while (!(RNG->SR & RNG_SR_DRDY));
                            ^
make[3]: *** [/home/travis/build/RIOT-OS/RIOT/examples/default/bin/nucleo-f401/periph/hwrng.o] Error 1
make[2]: *** [ALL--/home/travis/build/RIOT-OS/RIOT/cpu/stm32f4/periph] Error 2
make[1]: *** [ALL--/home/travis/build/RIOT-OS/RIOT/cpu/stm32f4] Error 2
make[1]: *** Waiting for unfinished jobs....
make: *** [all] Error 2
```

I guess it's from the removed `#if RANDOM_NUMOF` at the top of the stm32f4 rng driver?

haukepetersen commented:
- changed commit messages
- created `micro-ecc` path with `-w` option
- added guard in `cpu/stm32f4/periph/hwrng.c`, as not all F4 variants come with an on-chip RNG. This fixes the issue with the `nucleo-f401`

haukepetersen commented:
rebased again

gebart commented:
Travis seem to fail randomly on applying the micro-ecc patches "Makefile already exists in index"
There were also some other random errors in Travis, so I kicked it again.

```
pkg_micro-ecc:stm32f4discovery:
Building application "micro-ecc" for "stm32f4discovery" with MCU "stm32f4".
cd "/home/travis/build/RIOT-OS/RIOT/pkg/micro-ecc/micro-ecc" && git am "../0001-Add-RIOT-Makefile.patch" || { git am --abort; exit 1; }; cd "/home/travis/build/RIOT-OS/RIOT/pkg/micro-ecc/micro-ecc" && git am "../0002-Include-RIOT-Hardware-RNG-interface.patch" || { git am --abort; exit 1; }; cd "/home/travis/build/RIOT-OS/RIOT/pkg/micro-ecc/micro-ecc" && git am "../0003-Adapted-to-changed-random-interface.patch" || { git am --abort; exit 1; };
Applying: Add RIOT Makefile
error: Makefile: already exists in index
Patch failed at 0001 Add RIOT Makefile
The copy of the patch that failed is found in:
   /home/travis/build/RIOT-OS/RIOT/pkg/micro-ecc/micro-ecc/.git/rebase-apply/patch
When you have resolved this problem, run "git am --continue".
If you prefer to skip this patch, run "git am --skip" instead.
To restore the original branch and stop patching, run "git am --abort".
make[1]: *** [/home/travis/build/RIOT-OS/RIOT/pkg/micro-ecc/micro-ecc/Makefile] Error 1
make: *** [/home/travis/build/RIOT-OS/RIOT/tests/pkg_micro-ecc/bin/stm32f4discovery/micro-ecc.a] Error 2
```

gebart commented:
needs rebase after #4767 was merged.
There's something wrong with the new patch to micro-ecc, I can't apply it on my local checkout.

gebart commented:
maybe we should just squash the patch for micro-ecc into a single file and be done, what's your opinion @haukepetersen?

haukepetersen commented:
rebased for now, will look at the patch file once I am in the office. I guess squashing all patches sounds reasonable to me.

haukepetersen commented:
fuck these patches, I can't get them to work anymore... This is annoying to spend time on this crap

gebart commented:
On Wed, Feb 10, 2016 at 10:47 AM, Hauke Petersen
<notifications@github.com> wrote:
>
> fuck these patches, I can't get them to work anymore... This is annoying to spend time on this crap

Agreed! let's see if this works... I have attached a fixed patch file
to this email.
>
> —
> Reply to this email directly or view it on GitHub.

haukepetersen commented:
nope, does still not build...

haukepetersen commented:
```
"make" -C /home/hauke/dev/riot/RIOT/pkg/micro-ecc
test -d micro-ecc || \
	git clone https://github.com/kmackay/micro-ecc.git /home/hauke/dev/riot/RIOT/pkg/micro-ecc/micro-ecc && \
		cd /home/hauke/dev/riot/RIOT/pkg/micro-ecc/micro-ecc && git reset --hard f3d46f0fb77b42535168a17b4b51802beb124a32
HEAD is now at f3d46f0 Fix Thumb-2 inline asm (#15)
cd "/home/hauke/dev/riot/RIOT/pkg/micro-ecc/micro-ecc" && git am "../0001-Add-RIOT-Makefile.patch" || { git am --abort; exit 1; }; cd "/home/hauke/dev/riot/RIOT/pkg/micro-ecc/micro-ecc" && git am "../0002-Include-RIOT-Hardware-RNG-interface.patch" || { git am --abort; exit 1; }; cd "/home/hauke/dev/riot/RIOT/pkg/micro-ecc/micro-ecc" && git am "../0003-Adapted-to-changed-random-interface.patch" || { git am --abort; exit 1; };
previous rebase directory /home/hauke/dev/riot/RIOT/pkg/micro-ecc/micro-ecc/.git/rebase-apply still exists but mbox given.
make[1]: *** [/home/hauke/dev/riot/RIOT/pkg/micro-ecc/micro-ecc/Makefile] Error 1
make: *** [/home/hauke/dev/riot/RIOT/tests/pkg_micro-ecc/bin/native/micro-ecc.a] Error 2
```

gebart commented:
@haukepetersen I probably messed up the patch...

haukepetersen commented:
sounds familiar :-)

gebart commented:
@haukepetersen can you try importing these files? https://gist.github.com/gebart/85e68a0382bf6cb1e1cf

gebart commented:
related: #4788

haukepetersen commented:
no luck
```
"make" -C /home/hauke/dev/riot/RIOT/pkg/micro-ecc
test -d micro-ecc || \
	git clone https://github.com/kmackay/micro-ecc.git /home/hauke/dev/riot/RIOT/pkg/micro-ecc/micro-ecc && \
		cd /home/hauke/dev/riot/RIOT/pkg/micro-ecc/micro-ecc && git reset --hard f3d46f0fb77b42535168a17b4b51802beb124a32
HEAD is now at f3d46f0 Fix Thumb-2 inline asm (#15)
cd "/home/hauke/dev/riot/RIOT/pkg/micro-ecc/micro-ecc" && git am "../0001-Add-RIOT-Makefile.patch" || { git am --abort; exit 1; }; cd "/home/hauke/dev/riot/RIOT/pkg/micro-ecc/micro-ecc" && git am "../0002-Include-RIOT-Hardware-RNG-interface.patch" || { git am --abort; exit 1; };
Applying: Add RIOT Makefile
Applying: Include RIOT Hardware RNG interface
error: patch failed: uECC.c:322
error: uECC.c: patch does not apply
error: patch failed: uECC.h:57
error: uECC.h: patch does not apply
Patch failed at 0001 Include RIOT Hardware RNG interface
The copy of the patch that failed is found in:
   /home/hauke/dev/riot/RIOT/pkg/micro-ecc/micro-ecc/.git/rebase-apply/patch
When you have resolved this problem, run "git am --continue".
If you prefer to skip this patch, run "git am --skip" instead.
To restore the original branch and stop patching, run "git am --abort".
make[1]: *** [/home/hauke/dev/riot/RIOT/pkg/micro-ecc/micro-ecc/Makefile] Error 1
make: *** [/home/hauke/dev/riot/RIOT/tests/pkg_micro-ecc/bin/airfy-beacon/micro-ecc.a] Error 2
```

haukepetersen commented:
rebased and pushed those adapted patch files anyway to keep this branch clean...

haukepetersen commented:
ok, now it works, but it's not pretty. I squashed my changes into the 0002xxx patch, but it only worked without the `-w` option and by re-creating both path files...

gebart commented:
@haukepetersen did you possibly copy+paste the patch files from the gist above using your editor which strips trailing whitespace? It won't apply if you strip the trailing spaces.
Also, you need to reset the micro-ecc working copy if you are going to try and apply a new patch, preferably rm -rf the whole working copy.

haukepetersen commented:
yes, my editor actually is removing white spaces, so that might caused trouble... I will try again

haukepetersen commented:
@gebart alright, that did the trick, now using your proposed patchfiles

gebart commented:
something is up with the static checks, the cppcheck message got cut off

```
entPtrArg): Assignment of function parameter has no effect outside the function.
```

gebart commented:
the full message is:
```
cpu/native/periph/hwrng.c:86: style (uselessAssignmentArg): Assignment of function parameter has no effect outside the function.
```

haukepetersen commented:
fixed, rebase, squashed

gebart commented:
Travis is green, let's merge this!

haukepetersen commented:
Awesome! @gebart thanks very much for your help!

kaspar030 commented:
Yeah, nice one!
